docs(ad-api): add motion api (#154)
autowarefoundation/autoware-documentation: docs/design/autoware-interfaces/ad-api/list/api/planning/index.md

# Planning API

- {{ link_ad_api('/api/planning/allow_start') }}
- {{ link_ad_api('/api/planning/state') }}
- {{ link_ad_api('/api/planning/factors') }}

## Description

This API manages the behavior that the vehicle plans.
Applications can notify the vehicle behavior to the people around and visualize it for operator and passengers.

## States

The planning state manages the stop and start of the vehicle.
Once the vehicle has stopped, the state will be STOPPED.
After this, when the vehicle tries to start (is still stopped), the state will be STARTING.
In this state, calling the start API changes the state to MOVING and the vehicle starts.
This mechanism can add processing such as announcements before the vehicle starts.
Depending on the configuration, the state may transition directly from STOPPED to MOVING.

![planning-state](./docs/state.drawio.svg)

| State    | Description                                     |
| -------- | ----------------------------------------------- |
| STOPPED  | The vehicle is stopped.                         |
| STARTING | The vehicle is stopped, but is trying to start. |
| MOVING   | The vehicle is moving.                          |

> ## Factors

Reviewer: I feel we need a bit more discussions about the design of this API.

The planning factors is an array of information on the behavior that the vehicle plans.

> Each factor has type shown below, pose, distance from the vehicle head to that pose, status, and detailed data depending on its type.

Reviewer 2: ```suggestion
Each factor has a type (shown below), pose, distance from the vehicle head to that pose, status, and detailed data depending on its type.
```

See the description of each behavior below for pose and status. And see the description of factor types for detailed data.

![planning-factors](./docs/factors.drawio.svg)

| Factor Type                 | Behavior         | Description                                            |
| --------------------------- | ---------------- | ------------------------------------------------------ |

> | SURROUND_OBSTACLE           | immediate stop   | There are obstacles nearby.                            |

Reviewer 2: What is the difference between a `SURROUND_OBSTACLE` and a `FRONT_OBSTACLE`? `surround` means all around, and so this word also covers obstacles in front and behind.

Author: cc: @kenji-miyake 
The front has a long range because it detects on the route. So is the `ROUTE_OBSTACLE` better?

Reviewer 2: I think this is a good name. The description can also be updated to say something like `There are obstacles along the route ahead` to make it clear that the obstacle is further ahead of the vehicle along its planned route. 

Then the description for `SURROUND_OBSTACLE` (`SURROUNDING_OBSTACLE` may be better 🤔 ) can be updated to say `There are obstacles immediately around the vehicle` (assuming this is a close-range check).

Author: Done https://github.com/autowarefoundation/autoware-documentation/pull/154/commits/70b0c863477c900ab713e0c702087186e96c0f77.

| FRONT_OBSTACLE              | planned stop     | There are obstacles ahead.                             |
| INTERSECTION                | planned stop     | There are obstacles in other lanes in the path.        |
| CROSSWALK                   | planned stop     | There are obstacles on the crosswalk.                  |

> | REAR_CHECK                  | planned stop     | There are obstacles behind like shoulder checks.       |

Reviewer 2: This description suggests that a shoulder check is an obstacle, which is not the case. I'm not sure of a good alternative, but here's one for now.

```suggestion
| REAR_CHECK                  | planned stop     | There are obstacles behind that would be in a human driver's blind spot.       |
```

| USER_DEFINED_DETECTION_AREA | planned stop     | There are obstacles in the predefined detection area.  |
| NO_STOPPING_AREA            | planned stop     | There is not enough space beyond the no stopping area. |
| STOP_SIGN                   | planned stop     | A stop by a stop sign.                                 |
| TRAFFIC_SIGNAL              | planned stop     | A stop by a traffic signal.                            |
| V2I_GATE_CONTROL            | planned stop     | A stop by a V2I gate.                                  |
| FROM_PRIVATE_ROAD           | planned stop     | A stop before merging from private area.               |
| SIDEWALK                    | planned stop     | A stop before crossing the sidewalk.                   |
| LANE_CHANGE                 | lane change      | A lane change.                                         |

> | AVOIDANCE1                  | lane change      | A lane change to avoid obstacles when avoiding.        |
> | AVOIDANCE2                  | lane change      | A lane change to avoid obstacles when returning.       |

Reviewer 2: If the aim is for the factor names to be understandable, we need to change AVOIDANCE1 and AVOIDANCE2 to be more clear. In the meantime, I have suggested alternative descriptions for these two factors

```suggestion
| AVOIDANCE1                  | lane change      | A lane change to avoid an obstacle in the current lane.        |
| AVOIDANCE2                  | lane change      | A lane change to return to the original lane after avoiding an obstacle.       |
```

> | AVOIDANCE2                  | lane change      | A lane change to return to the original lane after avoiding an obstacle. |

Author: Replaced with `AVOIDANCE_PATH_CHANGE` and `AVOIDANCE_PATH_RETURN`.

| DIRECTION_CHANGE            | direction change | A direction change such as turning left or right.      |
| EMERGENCY_STOP_OPERATION    | T.B.D.           | A stop by emergency instruction from the operator.     |

### Planned stop

The planned stop is a ordinary and predictable stop at the target position.

> As the vehicle approaches the stop position, this factor appears with status APPROACHING.

Reviewer 2: ```suggestion
As the vehicle approaches the stop position, this factor appears with a status of APPROACHING.
```

> | DIRECTION_CHANGE            | direction change | A direction change such as turning left or right.                        |

Reviewer: Mm... I personally felt that adding `DIRECTION_CHANGE` here is a bit unnatural, but I have no good alternative idea. :pleading_face:

Author: Replaced with `INTERSECTION`.

And when the vehicle reaches that position and stops, the status will be STOPPED.
The pose always indicates the stop position.

![planned-stop-factor](./docs/factors-planned-stop.drawio.svg)

### Immediate stop

> The immediate stop does not have the specified stop position.

Reviewer 2: ```suggestion
The immediate stop does not have a specified stop position.
```

> Usually the vehicle stop as soon as possible under permitted conditions.

Reviewer 2: ```suggestion
Usually the vehicle stop as soon as possible under the permitted conditions.
```

> This factor will appear with status STOPPING when a stop is needed.

Reviewer 2: ```suggestion
This factor will appear with a status of STOPPING when a stop is needed.
```

And when the vehicle stops, the status will be STOPPED.
The pose indicates the vehicle head or the predicted stop position if possible.

![immediate-stop-factor](./docs/factors-immediate-stop.drawio.svg)

### Direction change

> The direction change is a steering that requires turn indicators such as turning left or right.

Reviewer 2: ```suggestion
The direction change is a maneuver that requires use of turn indicators, such as when turning left or right.
```

As the vehicle approaches the position to start steering, this factor appears with status APPROACHING.
And when the vehicle reaches that position, the status will be ACTIVATED.
The pose indicates the start position when APPROACHING and the end position when ACTIVATED.

![direction-change-factor](./docs/factors-direction-change.drawio.svg)

### Lane change

The lane change is basically the same as the direction change.
But the vehicle will start steering at any position in the range depending on the situation.

> As the vehicle approaches the start position of the range, this factor appears with status APPROACHING.

Reviewer 2: ```suggestion
As the vehicle approaches the start position of the range, this factor appears with a status of APPROACHING.
```
